c_src/membrane_audio_mix_plugin/mixer.c

```diff
@@ -1,210 +1,246 @@
-        }
-    }
-
-    if (min < state->sample_min)
-    {
-        double quotient = min / (state->sample_min * 1.0);
-        scale(values, values_length, quotient, state);
-    }
-    else if (max > state->sample_max)
-    {
-        double quotient = max / (state->sample_max * 1.0);
-        scale(values, values_length, quotient, state);
+    if (state->queue[i] > max) {
+      max = state->queue[i];
     }
+  }
 
-    int samples_start = 0;
-    for (unsigned int i = 0; i < state->queue_length; ++i, samples_start += state->sample_size)
-    {
-        value_to_sample(state->queue[i], samples + samples_start, state->caps);
+  for (uint32_t i = 0; i < values_length; ++i) {
+    if (values[i] < min) {
+      min = values[i];
     }
 
-    for (unsigned int i = 0; i < values_length; ++i, samples_start += state->sample_size)
-    {
-        value_to_sample(values[i], samples + samples_start, state->caps);
+    if (values[i] > max) {
+      max = values[i];
     }
-
-    unifex_free(state->queue);
-    state->queue = NULL;
-    state->queue_length = 0;
+  }
+
+  if (min < state->sample_min) {
+    double quotient = (state->sample_min * 1.0) / min;
+    scale(values, values_length, quotient, state);
+  } else if (max > state->sample_max) {
+    double quotient = (state->sample_max * 1.0) / max;
+    scale(values, values_length, quotient, state);
+  }
+
+  int32_t samples_size = 0;
+  for (uint32_t i = 0; i < state->queue_length;
+       ++i, samples_size += state->sample_size) {
+    caps_audio_raw_value_to_sample(state->queue[i], samples + samples_size,
+                                   state->caps);
+  }
+
+  for (uint32_t i = 0; i < values_length;
+       ++i, samples_size += state->sample_size) {
+    caps_audio_raw_value_to_sample(values[i], samples + samples_size,
+                                   state->caps);
+  }
+
+  unifex_free(state->queue);
+  state->queue = NULL;
+  state->queue_length = 0;
 }
 
-void add_values(int64_t *values, unsigned int values_length, unsigned char *samples, unsigned int *samples_size, bool is_last_wave, UnifexState *state)
-{
-    bool is_wave_positive = state->is_wave_positive;
-    unsigned int start = 0;
-    unsigned int end = split_values(values, values_length, is_wave_positive);
-    *samples_size = (values_length + state->queue_length) * state->sample_size;
-    samples = unifex_alloc(*samples_size);
-    unsigned int samples_start = 0;
-    while (end < values_length)
-    {
-        get_samples(samples + samples_start, values + start, end - start, state);
-        samples_start += (end - start + state->queue_length) * state->sample_size;
-        start = end;
-        state->is_wave_positive = !is_wave_positive;
-        end = split_values(values, values_length, is_wave_positive);
-    }
+/**
+ * Cuts given values on every sign change and convert it to samples.
+ *
+ * Any remaining values are put into state's queue.
+ */
+void cut_values(int64_t *values, uint32_t values_length, uint8_t *samples,
```

Comment:
Reading this doc and function name I'm still not sure what it does. What does `cutting` mean here?

Reply:
I rewrote this doc, hope the purpose of this function is more clear. I still don't have better name for this function. `Cutting` means  dividing into parts which are converted to the samples separately.